releasenotes/notes/pluggable-high-level-synthesis-3af9976b22e012d9.yaml

```diff
@@ -0,0 +1,34 @@
+---
+features:
+  - |
+    Introduced a new high level synthesis plugin interface which is used to enable
+    using alternative synthesis techniques included in external packages
+    seamlessly with the :class:`~qiskit.transpiler.passes.HighLevelSynthesis`
+    transpiler pass. These alternative synthesis techniques can be specified for
+    any "higher-level" objects of type :class:`~.Operation`, as for example for
+    :class:`~.Clifford` and :class:`~.LinearFunction` objects. This plugin interface
+    is similar to the one for unitary synthesis. In the latter case, the details on writing
+    a new plugin appear in the :mod:`qiskit.transpiler.passes.synthesis.plugin` module documentation.
```

Comment:
Are the processes for writing a `UnitarySynthesisPlugin` and `HighLevelSynthesisPlugin` sufficiently similar that the documentation for unitary synthesis covers everything needed to write a high-level synthesis plugin? If so, would it make sense to generalize the text there to cover both (and pave the way for their eventual coming together)?

Reply:
Yes, indeed, writing a `HighLevelSynthesisPlugin` should be very similar to writing a `UnitarySynthesisPlugin` (and actually a bit simpler right now as the API for `HighLevelSynthesisPlugin` currently does not have any `support_*` functions -- but this may evolve in the future). This is also why I chose to refer to the unitary synthesis plugin documentation instead of copying and adapting it for high level synthesis. But the suggestion to generalize the text to cover both makes good sense, @mtreinish what do you think?

Reply:
I think we can add sections to the `qiskit.transpiler.passes.synthesis.plugin` docstring to integrate the high level synthesis as part of the module doc. I intended that module to be a place that supported documenting all synthesis plugin interfaces. At the time (1H 2021 when we added the unitary synthesis plugin interface) I was assuming we'd have separate ones for each type of synthesis (like adding a new clifford synthesis plugin interface, etc). So having details on high level synthesis added to the documentation there makes sense to me

Reply:
Thanks @mtreinish. Based on your and @kdk's comments, I have both improved the documentation of higher level synthesis plugin, and added it as a section in `qiskit.transpiler.passes.synthesis.plugin` documentation. This is essentially done in [bda0ac4](https://github.com/Qiskit/qiskit-terra/pull/8548/commits/bda0ac4be19ad2f0ac1487b9e24e3631883e5222), with a few minor improvements/fixes in later commits.